src/main/java/com/artipie/asto/s3/EstimatedContentCompliment.java

```diff
@@ -0,0 +1,164 @@
+
+/**
+ * Complements {@link Content} with size if size is unknown.
+ * <p>
+ * Size calculated by reading up to `limit` content bytes.
+ * If end of content has not been reached by reading `limit` of bytes
+ * then original content is returned.
+ *
+ * @since 0.34
+ */
+final class EstimatedContentCompliment {
+    /**
+     * The original content.
+     */
+    private final Content original;
+
+    /**
+     * The limit.
+     */
+    private final long limit;
+
+    /**
+     * Ctor.
+     *
+     * @param original Original content.
+     * @param limit Content reading limit.
+     */
+    EstimatedContentCompliment(final Content original, final long limit) {
+        this.original = original;
+        this.limit = limit;
+    }
+
+    /**
+     * Ctor.
+     *
+     * @param original Original content.
+     */
+    EstimatedContentCompliment(final Content original) {
+        this(original, Long.MAX_VALUE);
+    }
+
+    /**
+     * Initialize future of Content.
+     *
+     * @return The future.
+     */
+    public CompletionStage<Content> estimate() {
+        final CompletableFuture<Content> res;
+        if (this.original.size().isPresent()) {
+            res = CompletableFuture.completedFuture(this.original);
+        } else {
+            res = this.readUntilLimit();
+        }
+        return res;
+    }
+
+    /**
+     * Read until limit.
+     *
+     * @return The future.
+     */
+    private CompletableFuture<Content> readUntilLimit() {
+        final Path temp;
+        try {
+            temp = Files.createTempFile(
+                S3Storage.class.getSimpleName(),
+                ".upload.tmp"
+            );
+        } catch (final IOException ex) {
+            throw new UncheckedIOException(ex);
+        }
+        return new File(temp)
+            .write(this.original)
+            .thenCompose(
+                nothing ->
+                    Flowable.fromPublisher(
+                        new File(temp).content()
+                    )
+                        .map(Buffer::remaining)
+                        .scanWith(() -> 0L, Long::sum)
+                        .takeUntil(total -> total >= this.limit)
+                        .lastOrError()
+                        .to(SingleInterop.get())
+                        .toCompletableFuture()
+            )
+            .thenApply(
+                last -> {
+                    final Optional<Long> size;
+                    if (last >= this.limit) {
+                        size = Optional.empty();
+                    } else {
+                        size = Optional.of(last);
+                    }
+                    return size;
+                }
+            ).thenApply(
+                sizeOpt -> {
+                    final Flowable<ByteBuffer> data = Flowable.fromPublisher(
+                        new File(temp).content()
+                    ).doAfterTerminate(
+                        () -> Files.delete(temp)
+                    );
```

Comment:
@andreoss it seems we have the similar logic for file removing in two places. I'd remove this `afterTerminate` callback and remove the file in `handle`

Reply:
@g4s8 Removing file in `handle` does not work. The file is removed too soon causing `Early EOF` exception